### PR TITLE
Create conan package recipe

### PR DESCRIPTION
**Pull request by spjuanjoc**
- Add a conan recipe to test locally the usage of SOCI as a conan package.
- I will need some help to add the rest of the back-ends, since I have only been able to test the sqlite3.
- This recipe is not added to the conan center yet.

### CODE REVIEW

**Comment by vadz** [comment]
Thanks for the PR!

Unfortunately I know even less about Conan than about CMake, so I can't really say anything about this. Are there any other Conan users here? Any comments?

Ideal would be to avoid having Conan-only examples and add an example to SOCI and build it normally.

Also, I'd rather not apply the changes to `.gitignore`, but this is a detail.

**Comment by spjuanjoc** [comment]
Thanks for the review. 
What do you mean with conan-only examples? The example project is about how SOCI can be included in a CMake project using Conan, making it easier to handle the dependencies: Conan does the compilation. However, the only example I have is using `slqite3` as backend.

Should I remove the changes to the `.gitignore` file?

Any clue about why the tests in the Travis Jobs 6 and 10 failed? I didn't modify any other project file than the `.gitignore`
The following tests failed:
- soci_odbc_test_postgresql (Failed)
- soci_odbc_test_postgresql_static (Failed)
- soci_oracle_test (Failed)
- soci_oracle_test_static (Failed)
Errors while running CTest

**Comment by vadz** [comment]
> What do you mean with conan-only examples? 

Example which is not used in the normal SOCI build and so not tested by the CI builds and so risk getting broken without anybody noticing it in the future. I'd rather add some new `examples/hello` to contain this simple example and then use/reference it from Conan-specific files.

> Should I remove the changes to the `.gitignore` file?

Yes, please.
 
> Any clue about why the tests in the Travis Jobs 6 and 10 failed? I didn't modify any other project file than the `.gitignore`
> The following tests failed:
> 
>     * soci_odbc_test_postgresql (Failed)
> 
>     * soci_odbc_test_postgresql_static (Failed)

These (spurious) errors should be fixed now in master and so shouldn't happen any more.
 
>     * soci_oracle_test (Failed)
> 
>     * soci_oracle_test_static (Failed)
>       Errors while running CTest

Unfortunately Oracle test is flaky on Travis infrastructure and I have no idea why, even though I spent many hours on it already (and @mloskot had probably spent even more before). Please just ignore this one.

**Comment by mloskot** [comment]
@vadz @spjuanjoc Here is more on my story with Oracle on Travis CI with links to even more details
- https://github.com/SOCI/soci/pull/326
- https://github.com/SOCI/soci/pull/579
- https://github.com/SOCI/soci/pull/727


Basically, the issue was due to these:
- Oracle packages for Ubuntu/Debian available on https://oss.oracle.com/debian/ have rotted without any updates
- Oracle requires registration and sign-in form to let you download Oracle XE binaries which obviously was/is a pain in the ass obstacle for any unsupervised downloads by CI jobs
- At that time, there was no docker image with Oracle (not sure if it's available now, but there were issues with Oracle calling for removal of images from Docker hub)

File: .conan/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+    def requirements(self):
+        if self.options.sqlite3:
+            self.requires("sqlite3/3.33.0")
+        # and so on for the rest of backends
+        # ToDo add the dependencies for the missing backends
+
+    def source(self):
+        self.run("git clone https://github.com/SOCI/soci.git")
```
[the remark this answers is not in the document]
Comment (spjuanjoc) [reply .conan/conanfile.py]: Update to support multiple versions, starting from v 4.0.0

[the remark this answers is not in the document]
Comment (mloskot) [reply .conan/conanfile.py]: @spjuanjoc FYI, unless the current maintainers have changed that (/cc @vadz), only `master` and `release/4.0` branches are actively maintained, so the branch `release/3.2` may not receive any bug fixes.

**Comment by spjuanjoc** [comment]
I could use some help to identify the following information:
The options are related to those in the CMakeLists.txt.
For the dependencies, I'm looking for the library name and its supported version in SOCI, or if possible: the conan recipe.


|Option | Library generated | has dependencies | 
|:---: | --- | --- |
|shared | soci_core | no |
|empty | soci_empty | no |
|static | ?? | ?? |
|cxx11 | -- | y: `cxx_std > 11` |
|sqlite3 | soci_sqlite | y: `sqlite3/3.33.0` |
|db2 | ?? | yes |
|odbc | ?? | yes |
|oracle | ?? | yes |
|firebird | ?? | yes |
|mysql | ?? | yes |
|postgresql | ?? | yes |
|tests | ?? | ?? |

**Comment by vadz** [comment]
> @vadz @spjuanjoc Here is more on my story with Oracle on Travis CI with links to even more details

@mloskot Just for the record (and/if you're still somehow interested in using Oracle with Travis CI, maybe with your other projects): the current problem are the intermittent (but very, very frequent) "ORA-12520: TNS:listener could not find available handler for requested type of server" errors in the CI builds. I've spent quite some time trying to reproduce it locally or to somehow debug it on Travis but without anything to show for it.

I've also tried migrating to running Oracle XE 18, for which official Docker containers are available now, see [my branch](https://github.com/vadz/soci/tree/oracle-xe-18) but for some reason never finished this. I should probably return to this because the current situation is just untenable, I often need to restart the CI builds several times before they finally pass.

**Comment by vadz** [comment]
@spjuanjoc For the library names you can just build SOCI using the normal CMake-based build system and look at them, but they're all pretty consistent and as you'd expect them to be anyhow, I'm not sure what exactly is your question?

**Comment by spjuanjoc** [comment]
@vadz I have tried building all the back ends, however, since I don't have all the dependencies libraries installed I get the message:
`WARNING: DB2 libraries not found, some features will be disabled. ` and so on.
I do have used  `sqlite3` so I know the library I can use is `libsqlite3-0`
It would help to know the other lib names for the backends to put them in the dependencies column (3).
For example in the SOCI documentation I found DB2: db2profile, should I say this is `libdb2profile`?
Also, I see that SOCI generates, when enabled, `libsoci_sqlite3`. Is this done in a similar way for the other backends? That's the column 2.

**Comment by vadz** [comment]
You can look at `scripts/travis/before_install_xxx.sh` scripts to check what needs to be installed for each backend. For DB/2 there is `db2exc` package available in an extra repository. The example of using `db2profile` in the docs is not very clear, but this is just a (shell) profile file installed as part of DB/2, i.e. not a dependency or whatever.

For the column 2, this is indeed similar for the other backends. Most of them can be built to confirm it and, if you'd like, you can run `ldd` on the libraries to compile an authoritative list of their dependencies.

**Comment by mloskot** [comment]
@vadz Thanks for the explanation. Perhaps something has changed w/ Travis CI images networking, no idea really.
Perhaps switching to GitHub Actions would improve it, no idea either.

**Comment by vadz** [comment]
Looking at the latest version, it seems that it still adds Conan-specific tests instead of building the normal unit tests with Conan and I still don't think it's a good idea -- these tests risk getting out of date, as we will almost certainly forget to update them if the API changes and they're not tested by the CI builds.

I.e. I don't think it's a good idea to merge this in the current state, sorry. Please reopen if you can

1. Reuse the existing tests.
2. Add other backends.
3. Add a job testing this to GitHub CI workflow.

TIA!

**Comment by spjuanjoc** [comment]
@vadz Yes, you're right. I opened this PR to get some guidance about how to include soci in conan but then I realized it was not required to include the conan code in this repo. I apologize, I forgot to mention that I worked on the Conan recipe and it was merged recently, so now SOCI/4.0.1 is available to install via conan using the backends sqlite3, odbc, mysql, and postgresql. The recipe can be found in the Conan Center Index: [CCI](https://conan.io/center/soci)

**Comment by vadz** [comment]
Great, thanks!

Pity that Firebird is missing (DB/2 and Oracle are not open source, so this is quite different), but it's still nice that all the other ones are available, thanks for your work!

Also, could you please check if nothing got broken with the latest 4.0 branch? I plan to release 4.0.2 on Monday and would appreciate any testing. TIA!

**Comment by spjuanjoc** [comment]
That's some great news! Sure, I'll take a look at it and see if I can include the new version into the Conan recipe.
About Firebird, sadly it is not in conan yet, as far as I know. If it gets added then it would be quite simple to enable it in the current soci recipe.  
It would be nice if I can add the steps to install SOCI via Conan to the documentation but I would need some guidance, I'm not sure where to begin. I guess it is just by modifying the docs/installation.md, and then it gets updated in the sourceforge doc by the CI? If so it would add something like [this](https://spjuanjoc.github.io/avanzado/sql-incrustado-soci/#conan) I wrote (in english of course)

**Comment by vadz** [comment]
Yes, `docs/installation.md` is the right place to describe this, probably in a new section about Conan. TIA!

**Comment by spjuanjoc** [comment]
Well, the 4.0.2 tests I ran locally worked fine. 
There is something odd, it seems that some "Find*.cmake" files are being copied into the package lib/ directory, Is this happening to you?

**Pull request closed without merging**